File: android/src/nz/co/simplypayroll/charts/data/BaseDataSetProxy.java
package nz.co.simplypayroll.charts.data;

import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.BaseDataSet;
import com.github.mikephil.charting.utils.Utils;

import org.appcelerator.kroll.KrollDict;
import org.appcelerator.kroll.KrollProxy;
import org.appcelerator.kroll.annotations.Kroll;
import org.appcelerator.titanium.TiApplication;
import org.appcelerator.titanium.TiC;
import org.appcelerator.titanium.util.TiColorHelper;
import org.appcelerator.titanium.util.TiConvert;
import org.appcelerator.titanium.util.TiUIHelper;

import java.util.HashMap;
import java.util.Iterator;
import java.util.List;

import nz.co.simplypayroll.charts.utils.TiConverter;

/**
 * Created by adrienblot on 28/06/16.
 */
@Kroll.proxy
public class BaseDataSetProxy extends KrollProxy {

    private static final String LCAT = "BaseDataSetProxy";

    public BaseDataSet dataSet;

    public BaseDataSetProxy() {
        super();
        Utils.init(TiApplication.getInstance().getApplicationContext().getResources());
    }

    /**
     * Use this method to tell the data set that the underlying data has changed.
     */
    @Kroll.method
    public void notifyDataSetChanged() {
        dataSet.notifyDataSetChanged();
    }


    /**
     * ###### ###### COLOR GETTING RELATED METHODS ##### ######
     */
    @Kroll.getProperty @Kroll.method
    public String[] getColors() {
        if(hasProperty("colors")) {
            return (String[])getProperty("colors");
        } else {
            List<Integer> colors = dataSet.getColors();
            Iterator<Integer> iterator = colors.iterator();
            String[] colorsProperty = new String[colors.size()];
            int i = 0;
            while(iterator.hasNext()){
                Integer obj = iterator.next();
                colorsProperty[i] = Integer.toHexString(obj);
                i++;
            }
            setProperty("colors", colorsProperty);
            return colorsProperty;
        }
    }

    @Kroll.getProperty @Kroll.method
    public String[] getValueTextColors() {
        if(hasProperty("valueTextColors")) {
            return (String[])getProperty("valueTextColors");
        } else {
            List<Integer> colors = dataSet.getValueColors();
            Iterator<Integer> iterator = colors.iterator();
            String[] colorsProperty = new String[colors.size()];
            int i = 0;
            while(iterator.hasNext()){
                Integer obj = iterator.next();
                colorsProperty[i] = Integer.toHexString(obj);
                i++;
            }
            setProperty("valueTextColors", colorsProperty);
            return colorsProperty;
        }
    }

    @Kroll.method
    public String getColor(@Kroll.argument(optional=true) Object colorIndex) {
        int index = 0;
        if(colorIndex != null) {
            index = TiConvert.toInt(colorIndex, 0);
        }
        if(hasProperty("colors")) {
            String[] colors = (String[])getProperty("colors");
            if(colors.length == 0)
                return null;
            else
                return colors[index % colors.length];
        } else {
            return Integer.toHexString(dataSet.getColor(index));
        }
    }


    /**
     * ###### ###### COLOR SETTING RELATED METHODS ##### ######
     */

    /**
     * Sets the colors that should be used fore this DataSet. Colors are reused
     * as soon as the number of Entries the DataSet represents is higher than
     * the size of the colors array. If you are using colors from the resources,
     * make sure that the colors are already prepared (by calling
     * getResources().getColor(...)) before adding them to the DataSet.
     *
     * @param colors
     */
    @Kroll.setProperty @Kroll.method
    public void setColors(Object colors) {
        int [] colorsArray = null;
        String [] colorsProperty = null;
        if ((colors.getClass().isArray())) {
            colorsProperty = TiConvert.toStringArray((Object[])colors);

        } else {
            colorsProperty = new String[1];
            colorsProperty[0] = TiConvert.toString(colors);
        }

        if(colorsProperty != null) {
            colorsArray = TiConverter.toColorArray(colorsProperty);
            setProperty("colors", colorsProperty);
            dataSet.setColors(colorsArray);
        }
    }


    /**
     * Adds a new color to the colors array of the DataSet.
     *
     * @param color
     */
    @Kroll.method
    public void addColor(String color) {
        String [] newColorsProperty = null;
        if(hasProperty("colors")) {
            String [] colorsProperty = (String[])getProperty("colors");
            newColorsProperty = new String [colorsProperty.length + 1];
            for (int i = 0; i < colorsProperty.length; i++) {
                newColorsProperty[i] = colorsProperty[i];
            }
            newColorsProperty[newColorsProperty.length-1] = color;
        } else {
            List<Integer> colors = dataSet.getColors();
            Iterator<Integer> iterator = colors.iterator();
            newColorsProperty = new String[colors.size() + 1];
            int i = 0;
            while(iterator.hasNext()){
                Integer obj = iterator.next();
                newColorsProperty[i] = Integer.toHexString(obj);
                i++;
            }
            newColorsProperty[newColorsProperty.length-1] = color;

        }
        if(newColorsProperty != null)
            setProperty("colors", newColorsProperty);

        dataSet.addColor(TiConvert.toColor(color));
    }

    /** ###### ###### OTHER STYLING RELATED METHODS ##### ###### */

    @Kroll.setProperty @Kroll.method
    public void setLabel(String label) {
        dataSet.setLabel(label);
    }

    @Kroll.getProperty @Kroll.method
    public String getLabel() {
        return dataSet.getLabel();
    }

    @Kroll.setProperty @Kroll.method
    public void setHighlightEnabled(Object enabled) {
        dataSet.setHighlightEnabled(TiConvert.toBoolean(enabled, false));
    }

    @Kroll.getProperty @Kroll.method
    public boolean getHighlightEnabled() {
        return dataSet.isHighlightEnabled();
    }

    /*@Override
    public void setValueFormatter(ValueFormatter f) {

        if (f == null)
            return;
        else
            mValueFormatter = f;
    }*/

    /*@Override
    public ValueFormatter getValueFormatter() {
        if (mValueFormatter == null)
            return new DefaultValueFormatter(1);
        return mValueFormatter;
    }*/

    @Kroll.method
    public String getValueTextColor(@Kroll.argument(optional=true) Object colorIndex) {
        int index = 0;
        if(colorIndex != null) {
            index = TiConvert.toInt(colorIndex, 0);
        }
        if(hasProperty("valueTextColors")) {
            String[] colors = (String[])getProperty("valueTextColors");
            if(colors.length == 0)
                return null;
            else
                return colors[index % colors.length];
        } else {
            return Integer.toHexString(dataSet.getValueTextColor(index));
        }
    }


    /**
     * ###### ###### COLOR SETTING RELATED METHODS ##### ######
     */

    /**
     * Sets the colors that should be used fore this DataSet. Colors are reused
     * as soon as the number of Entries the DataSet represents is higher than
     * the size of the colors array. If you are using colors from the resources,
     * make sure that the colors are already prepared (by calling
     * getResources().getColor(...)) before adding them to the DataSet.
     *
     * @param colors
     */
    @Kroll.setProperty @Kroll.method
    public void setValueTextColors(Object colors) {
        List<Integer> colorsArray = null;
        String [] colorsProperty = null;
        if ((colors.getClass().isArray())) {
            colorsProperty = TiConvert.toStringArray((Object[])colors);

        } else {
            colorsProperty = new String[1];
            colorsProperty[0] = TiConvert.toString(colors);
        }

        if(colorsProperty != null) {
            colorsArray = TiConverter.toColorList(colorsProperty);
            setProperty("valueTextColors", colorsProperty);
            dataSet.setValueTextColors(colorsArray);
        }
    }

    @Kroll.setProperty @Kroll.method
    public void setValueTextFont(Object value) {
        if (value instanceof HashMap) {
            HashMap<String, Object> font = (HashMap<String, Object>) value;
            setProperty(TiC.PROPERTY_FONT, font);
            if (font.containsKey(TiC.PROPERTY_FONTFAMILY)) {
                dataSet.setValueTypeface(TiUIHelper.toTypeface(TiApplication.getInstance().getApplicationContext(), TiConvert.toString(font, TiC.PROPERTY_FONTFAMILY)));
            } else if (font.containsKey(TiC.PROPERTY_FONT_FAMILY)) {
                dataSet.setValueTypeface(TiUIHelper.toTypeface(TiApplication.getInstance().getApplicationContext(), TiConvert.toString(font, TiC.PROPERTY_FONT_FAMILY)));
            }
            if (font.containsKey(TiC.PROPERTY_FONTSIZE)) {
                dataSet.setValueTextSize(TiUIHelper.getSize(TiConvert.toString(font, TiC.PROPERTY_FONTSIZE)));
            } else if(font.containsKey(TiC.PROPERTY_FONT_SIZE)) {
                dataSet.setValueTextSize(TiUIHelper.getSize(TiConvert.toString(font, TiC.PROPERTY_FONT_SIZE)));
            }
        }
    }


    @Kroll.setProperty @Kroll.method
    public void setDrawValuesEnabled(Object enabled) {
        dataSet.setDrawValues(TiConvert.toBoolean(enabled, true));
    }

    @Kroll.getProperty @Kroll.method
    public boolean getDrawValuesEnabled() {
        return dataSet.isDrawValuesEnabled();
    }

    @Kroll.setProperty @Kroll.method
    public void setVisible(Object enabled) {
        dataSet.setVisible(TiConvert.toBoolean(enabled, true));
    }

    @Kroll.getProperty @Kroll.method
    public boolean getVisible() {
        return dataSet.isVisible();
    }

    @Kroll.getProperty @Kroll.method
    public String getAxisDependency() {
        return TiConverter.AxisDependencytoString(dataSet.getAxisDependency());
    }

    @Kroll.setProperty @Kroll.method
    public void setAxisDependency(String dependency) {
        dataSet.setAxisDependency(TiConverter.StringtoAxisDependency(dependency));
    }


    /** ###### ###### DATA RELATED METHODS ###### ###### */

    @Kroll.method
    public int getIndexInEntries(int xIndex) {
        return dataSet.getIndexInEntries(xIndex);
    }

    @Kroll.method
    public void removeFirst() {
        dataSet.removeFirst();
    }

    @Kroll.method
    public void removeLast() {
        dataSet.removeLast();
    }

    @Kroll.method
    public void removeEntry(int xIndex) {
        dataSet.removeEntry(xIndex);
    }

    /*
    @Override
    public boolean contains(T e) {

        for(int i = 0; i < getEntryCount(); i++) {
            if(getEntryForIndex(i).equals(e))
                return true;
        }

        return false;
    }*/
}

File: iphone/Classes/data/BaseDataSetProxy.m
//
//  BaseDataSetProxy.m
//  SimplyCharts
//
//  Created by Adrien Blot on 30/06/16.
//
//

#import <Foundation/Foundation.h>
#import "BaseDataSetProxy.h"

static int nextProxyId = 1;

@implementation BaseDataSetProxy

-(id)init
{
    // This is the designated initializer method and will always be called
    // when the proxy is created.
    
    // Generate a unique identifier so the user can see which proxy
    // instance is being created or destroyed (for demonstration purposes only).
    proxyId = nextProxyId++;
    
    NSLog(@"[PROXY LIFECYCLE EVENT] init with proxy id of %d", proxyId);
    
    return [super init];
}

-(id)initWithDataSet: (ChartBaseDataSet*)newdataSet
{
    dataSet = newdataSet;
    colors = [TiConverter toTiColorArray: [dataSet colors]];
    valueColors = [TiConverter toTiColorArray: [dataSet valueColors]];
    return [self init];
}

-(void)_destroy
{
    // This method is called from the dealloc method and is good place to
    // release any objects and memory that have been allocated for the proxy.
    RELEASE_TO_NIL(dataSet);
    [super _destroy];
}

-(ChartBaseDataSet*)getDataSet
{
    return dataSet;
}

-(id)notifyDataSetChanged
{
    [dataSet notifyDataSetChanged];
}


/**
 * ###### ###### COLOR GETTING RELATED METHODS ##### ######
 */

-(NSArray<TiColor *> *)Colors
{
    if(colors == nil) {
        colors = [TiConverter toTiColorArray: [dataSet colors]];
    }
    return colors;
}

-(NSArray<TiColor *> *) valueTextColors
{
    if(valueColors == nil) {
        valueColors = [TiConverter toTiColorArray: [dataSet valueColors]];
    }
    return valueColors;
}

-(TiColor*) getColor:(id)args
{
    ENSURE_ARRAY(args);
    int index = ([args count] > 0) ? [TiUtils intValue:[args objectAtIndex:0]] : 0;
    return colors[index % [colors count]];

}


/**
 * ###### ###### COLOR SETTING RELATED METHODS ##### ######
 */

/**
 * Sets the colors that should be used fore this DataSet. Colors are reused
 * as soon as the number of Entries the DataSet represents is higher than
 * the size of the colors array. If you are using colors from the resources,
 * make sure that the colors are already prepared (by calling
 * getResources().getColor(...)) before adding them to the DataSet.
 *
 * @param colors
 */
-(void)setColors:(id)newColors
{
    ENSURE_ARRAY(newColors);
    colors = [TiConverter toTiColorArray: (NSArray *)newColors];
    [dataSet setColors:[TiConverter toUIColorArray: colors]];
}


/**
 * Adds a new color to the colors array of the DataSet.
 *
 * @param color
 */
-(void)addColor:(id)args
{
    enum Args {
        kArgColor = 0,
        kArgCount
    };
    
    ENSURE_ARRAY(args);
    ENSURE_ARG_COUNT(args, kArgCount);
    TiColor *color = [TiUtils colorValue:[args objectAtIndex:kArgColor]];
    [colors addObject: color];
    [dataSet addColor: [color _color]];
}

/** ###### ###### OTHER STYLING RELATED METHODS ##### ###### */

-(void)setLabel:(id)label
{
    ENSURE_STRING(label);
    [dataSet setLabel:[TiUtils stringValue:label]];
}

-(NSString *)label
{
    return [dataSet label];
}

-(void)setHighlightEnabled:(id)enabled
{
    [dataSet setHighlightEnabled:[TiUtils boolValue:enabled]];
}

-(NSNumber *)highlightEnabled
{
    return NUMBOOL([dataSet highlightEnabled]);
}


/*@Override
 public void setValueFormatter(ValueFormatter f) {
 
 if (f == null)
 return;
 else
 mValueFormatter = f;
 }*/

/*@Override
 public ValueFormatter getValueFormatter() {
 if (mValueFormatter == null)
 return new DefaultValueFormatter(1);
 return mValueFormatter;
 }*/

-(TiColor*) getValueTextColor:(id)args
{
    ENSURE_ARRAY(args);
    int index = ([args count] > 0) ? [TiUtils intValue:[args objectAtIndex:0]] : 0;
    return valueColors[index % [valueColors count]];
    
}



/**
 * ###### ###### COLOR SETTING RELATED METHODS ##### ######
 */

/**
 * Sets the colors that should be used fore this DataSet. Colors are reused
 * as soon as the number of Entries the DataSet represents is higher than
 * the size of the colors array. If you are using colors from the resources,
 * make sure that the colors are already prepared (by calling
 * getResources().getColor(...)) before adding them to the DataSet.
 *
 * @param colors
 */
-(void)setValueTextColors:(id)newColors
{
    ENSURE_ARRAY(newColors);
    valueColors = [TiConverter toTiColorArray: (NSArray *)newColors];
    [dataSet setValueColors:[TiConverter toUIColorArray: colors]];
}

-(void)setValueTextFont: (id)value
{
    NSLog(@"[VIEW LIFECYCLE EVENT] setDescriptionFont_");
    ENSURE_DICT(value);
    NSDictionary *newdescriptionFont = (NSDictionary*)value;
    id fontFamily = [newdescriptionFont valueForKey: @"fontFamily"];
    UIFont *newFont = nil;
    CGFloat pointSize;
    if(fontFamily != nil) {
        newFont = [[TiUtils fontValue:fontFamily] font];
        [dataSet setValueFont: newFont];
    } else {
        newFont =  [dataSet valueFont];
    }
    
    if(newFont != nil) {
        id fontSize = [newdescriptionFont valueForKey: @"fontSize"];
        if(fontSize != nil) {
            pointSize = [TiUtils sizeValue:fontSize];
            newFont = [newFont fontWithSize: pointSize];
        }
        [dataSet setValueFont: newFont];
    }
}

-(NSDictionary*)valueTextFont
{
    UIFont* font = [dataSet valueFont];
    NSDictionary* fontDict = [[NSDictionary alloc] initWithObjectsAndKeys: [font fontName], @"fontFamily", NUMDOUBLE([font pointSize]), @"fontSize", nil];
    return fontDict;
}


-(void)setDrawValuesEnabled:(id)enabled
{
    [dataSet setDrawValuesEnabled:[TiUtils boolValue:enabled]];
}

-(NSNumber *)drawValuesEnabled
{
    return NUMBOOL([dataSet drawValuesEnabled]);
}

-(void)setVisible:(id)enabled
{
    [dataSet setVisible:[TiUtils boolValue:enabled]];
}

-(NSNumber *)visible
{
    return NUMBOOL([dataSet visible]);
}


-(NSString *)axisDependency
{
    return [TiConverter axisDependencytoString: [dataSet axisDependency]];
}

-(void)setAxisDependency:(id)dependency
{
    ENSURE_STRING(dependency);
    [dataSet setAxisDependency:[TiConverter stringtoAxisDependency:[TiUtils stringValue:dependency]]];
}


/** ###### ###### DATA RELATED METHODS ###### ###### */

-(void)removeFirst:(id)args
{
    [dataSet removeFirst];
}

-(void)removeLast:(id)args
{
    [dataSet removeLast];
}

-(void)removeEntry:(id)args
{
    enum Args {
        kArgXIndex = 0,
        kArgCount
    };

    ENSURE_ARRAY(args);
    ENSURE_ARG_COUNT(args, kArgCount);
    int xIndex = [TiUtils intValue:[args objectAtIndex:kArgXIndex]];
    [dataSet removeEntry: xIndex];
}

/*
 @Override
 public boolean contains(T e) {
 
 for(int i = 0; i < getEntryCount(); i++) {
 if(getEntryForIndex(i).equals(e))
 return true;
 }
 
 return false;
 }*/


@end